File: WPF/Controls/ObjList.vb
Namespace Common.Controls

    Public Class ObjList
        Inherits BaseList(Of Obj)
        Implements Markup.IAddChild

        Public Sub New(ByVal Parent As TextBlock)
            Me._Parent = Parent
        End Sub

        Private Sub GotIn(ByVal Obj As Obj)
            Obj.ReportParent(Me.Parent)
            Me.Parent.ReportObjGotIn(Obj)
        End Sub

        Private Sub WentOut(ByVal Obj As Obj)
            Me.Parent.ReportObjWentOut(Obj)
        End Sub

        Public Overrides ReadOnly Property Count As Integer
            Get
                Return Me.List.Count
            End Get
        End Property

        Default Public Overrides Property Item(index As Integer) As Obj
            Get
                Return Me.List.Item(index)
            End Get
            Set(value As Obj)
                Me.WentOut(Me.List.Item(index))
                Me.List.Item(index) = value
                Me.GotIn(value)
                Me.Parent.ReportObjChanged()
            End Set
        End Property

        Public Overrides Sub Clear()
            For Each O In Me.List
                Me.WentOut(O)
            Next
            Me.List.Clear()
            Me.Parent.ReportObjChanged()
        End Sub

        Public Overrides Sub Insert(index As Integer, item As Obj)
            Me.List.Insert(index, item)
            Me.GotIn(item)
            Me.Parent.ReportObjChanged()
        End Sub

        Public Overrides Sub RemoveAt(index As Integer)
            Me.WentOut(Me.List.Item(index))
            Me.List.RemoveAt(index)
            Me.Parent.ReportObjChanged()
        End Sub

        Protected Overrides Function IEnumerable_1_GetEnumerator() As IEnumerator(Of Obj)
            Return Me.List.GetEnumerator()
        End Function

        Public Function GetEnumerator() As List(Of Obj).Enumerator
            Return Me.List.GetEnumerator()
        End Function

        Public Sub AddChild(value As Object) Implements Markup.IAddChild.AddChild
            Me.Add(DirectCast(value, Obj))
        End Sub

        Public Sub AddText(text As String) Implements Markup.IAddChild.AddText
            Throw New NotSupportedException()
        End Sub

#Region "Parent Property"
        Private ReadOnly _Parent As TextBlock

        Public ReadOnly Property Parent As TextBlock
            Get
                Return Me._Parent
            End Get
        End Property
#End Region

        Private ReadOnly List As List(Of Obj) = New List(Of Obj)()

    End Class

End Namespace

File: WPF/Controls/ObjList.cs
using System.Windows;
using System.Threading.Tasks;
using Mono;
using System.Data;
using System.Diagnostics;
using Microsoft.VisualBasic;
using System.Linq;
using System.Collections.Generic;
using System.Collections;
using System.Windows.Controls;
using System;
using System.Xml.Linq;

namespace Ks
{
    namespace Common.Controls
    {
        public class ObjList : BaseList<Obj>, System.Windows.Markup.IAddChild
        {
            public ObjList(TextBlock Parent)
            {
                this._Parent = Parent;
            }

            private void GotIn(Obj Obj)
            {
                Obj.ReportParent(this.Parent);
                this.Parent.ReportObjGotIn(Obj);
            }

            private void WentOut(Obj Obj)
            {
                this.Parent.ReportObjWentOut(Obj);
            }

            public override int Count
            {
                get
                {
                    return this.List.Count;
                }
            }

            public override Obj this[int index]
            {
                get
                {
                    return this.List[index];
                }
                set
                {
                    this.WentOut(this.List[index]);
                    this.List[index] = value;
                    this.GotIn(value);
                    this.Parent.ReportObjChanged();
                }
            }

            public override void Clear()
            {
                foreach (var O in this.List)
                    this.WentOut(O);
                this.List.Clear();
                this.Parent.ReportObjChanged();
            }

            public override void Insert(int index, Obj item)
            {
                this.List.Insert(index, item);
                this.GotIn(item);
                this.Parent.ReportObjChanged();
            }

            public override void RemoveAt(int index)
            {
                this.WentOut(this.List[index]);
                this.List.RemoveAt(index);
                this.Parent.ReportObjChanged();
            }

            protected override IEnumerator<Obj> _GetEnumerator()
            {
                return this.List.GetEnumerator();
            }

            public List<Obj>.Enumerator GetEnumerator()
            {
                return this.List.GetEnumerator();
            }

            public void AddChild(object value)
            {
                this.Add((Obj)value);
            }

            public void AddText(string text)
            {
                throw new NotSupportedException();
            }

            private readonly TextBlock _Parent;

            public TextBlock Parent
            {
                get
                {
                    return this._Parent;
                }
            }

            private readonly List<Obj> List = new List<Obj>();
        }
    }
}
